adminapi/request.py

```diff
@@ -106,21 +142,53 @@ def send_request(endpoint, get_params=None, post_params=None):
     return json.loads(response.read().decode())
 
 
-def _build_request(endpoint, auth_token, get_params, post_params):
+def _build_request(endpoint, get_params, post_params):
+    """Wrap request data in an urllib Request instance
+
+    Aside from preparing the get and post data for transport, this function
+    authenticates the request using either an auth token or ssh keys.
+
+    Returns an urllib Request.
+    """
     if post_params:
         post_data = json.dumps(post_params, default=json_encode_extra)
     else:
         post_data = None
 
     timestamp = int(time.time())
-    app_id = calc_app_id(auth_token)
-    security_token = calc_security_token(auth_token, timestamp, post_data)
     headers = {
         'Content-Encoding': 'application/x-json',
         'X-Timestamp': str(timestamp),
-        'X-Application': app_id,
-        'X-SecurityToken': security_token,
     }
+
+    if Settings.auth_key:
+        headers['X-PublicKeys'] = Settings.auth_key.get_base64()
```

Comment:
true story, there is really no need to send public keys

Reply:
.get_base64() returns the public part of the private key http://docs.paramiko.org/en/2.4/api/keys.html#paramiko.pkey.PKey.get_base64

Reply:
There is plenty of reason to send the public key. Without it we would have to verify every signature against every public key in serveradmin. That would perform terribly.

I also felt uncomfortable about sending public keys initially[0], but there is really no problem with this. Hence Emre asked me to not beat around the bush and just send the public keys and I agress. Even github exposes your public keys. [1]

Also as we have discussed there is nothing about this that enables a MITM attack.

[0] https://github.com/innogames/serveradmin/pull/62/commits/3137f2b7339162d39503179303efa033d700eaac
[1] https://github.com/HerrSpace.keys

Reply:
Is it just me or is it a poor naming from paramiko for this method? ^^

Reply:
They are pretty terrible. There internal API is pretty bad. Most object are not printable in py3 either since their str method returns bytes and so on..

Reply:
yeah this api is not really nice. but after yesterday's discussion i don't have anything against it now